tests/unit/test_diffsync.py

```diff
@@ -81,11 +81,126 @@ def test_diffsync_get_by_uids_with_no_data(generic_diffsync):
+    generic_diffsync.add(person)
+    assert person == generic_diffsync._data[modelname][uid]  # pylint: disable=protected-access
+    assert person == generic_diffsync.get(PersonA, "Mikhail Yohman")
+
+
+def test_diffsync_add_raises_already_exists_with_updated_object(generic_diffsync):
+    intf = Interface(device_name="device1", name="eth0")
     # A DiffSync can store arbitrary DiffSyncModel objects, even if it doesn't know about them at definition time.
-    generic_diffsync.add(generic_diffsync_model)
-    with pytest.raises(ObjectAlreadyExists):
-        generic_diffsync.add(generic_diffsync_model)
+    generic_diffsync.add(intf)
+    # Create new interface with same identifiers so it's technically the same object, but set additional attribute
+    new_intf = Interface(device_name="device1", name="eth0", interface_type="1000base-t")
+    with pytest.raises(ObjectAlreadyExists) as error:
+        generic_diffsync.add(new_intf)
+    error_model = error.value.args[1]
+    assert isinstance(error_model, DiffSyncModel)
+    assert new_intf is error_model
+
+
+def test_diffsync_get_or_instantiate_create_non_existent_object(generic_diffsync):
+    intf_identifiers = {"device_name": "device1", "name": "eth1"}
+    intf = Interface(**intf_identifiers)
+
+    # Assert that the object does not currently exist.
+    with pytest.raises(ObjectNotFound):
+        generic_diffsync.get(Interface, intf_identifiers)
+
+    obj, created = generic_diffsync.get_or_instantiate(Interface, intf_identifiers)
+    assert obj == intf
+    assert created
+    assert obj is generic_diffsync.get(Interface, intf_identifiers)
+
+
+def test_diffsync_get_or_instantiate_retrieve_existing_object(generic_diffsync):
+    intf_identifiers = {"device_name": "device1", "name": "eth1"}
+    intf = Interface(**intf_identifiers)
+    generic_diffsync.add(intf)
+
+    obj, created = generic_diffsync.get_or_instantiate(Interface, intf_identifiers)
+    assert obj is intf
+    assert not created
+
+
+def test_diffsync_get_or_instantiate_retrieve_create_non_existent_w_attrs(generic_diffsync):
+    intf_identifiers = {"device_name": "device1", "name": "eth1"}
+    intf_attrs = {"interface_type": "1000base-t", "description": "Testing"}
+
+    obj, created = generic_diffsync.get_or_instantiate(Interface, intf_identifiers, intf_attrs)
+    assert created
+    assert obj.interface_type == "1000base-t"
+    assert obj.description == "Testing"
+    assert obj is generic_diffsync.get(Interface, intf_identifiers)
+
+
+def test_diffsync_get_or_instantiate_retrieve_existing_object_wo_attrs(generic_diffsync):
+    intf_identifiers = {"device_name": "device1", "name": "eth1"}
+    intf = Interface(**intf_identifiers)
+    generic_diffsync.add(intf)
+
+    obj, created = generic_diffsync.get_or_instantiate(Interface, intf_identifiers)
+    assert obj is intf
+    assert not created
+    assert obj.interface_type == "ethernet"
+    assert obj.description is None
```

Comment:
Would it make sense to also have a test case where you have an existing object, then call `get_or_instantiate` with matching `ids` but with `attrs` that differ from the existing object, and assert that the retrieved object preserves the original object's `attrs`?